### PR TITLE
Add IPv6 test cases to kube-proxy server test.

### LINKED ISSUE #47313: Need IPv6 test cases for kube-proxy app server
**Is this a request for help?**  No.

**What keywords did you search in Kubernetes issues before filing this one?** is:issue is:open ipv6

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

**Kubernetes version** (use `kubectl version`): N/A, these are unit test changes in latest master

**Environment**:
- **Cloud provider or hardware configuration**: Kubernetes container dev env on Ubuntu 16.04.1
- **OS** (e.g. from /etc/os-release): Ubuntu 16.04.1
- **Kernel** (e.g. `uname -a`): 4.4.0-78-generic #99-Ubuntu x86_64 GNU/Linux 
- **Install tools**:
- **Others**:

**What happened**:
Upon inspection, no IPv6 test cases found in kubernetes/cmd/kube-proxy/app/server_test.go.

**What you expected to happen**:
Expect to see some IPv6 test cases exercised for kube-proxy server code
(kubernetes/cmd/kube-proxy/app/server.go). Also need some code coverage
for negative conditions in server_test.go.

**How to reproduce it** (as minimally and precisely as possible):
Run unit tests in kubernetes/cmd/kube-proxy/app/server_test.go

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by leblancd**
**What this PR does / why we need it**: This change adds some IPv6 test cases for the kube-proxy server. Also adds some test cases for negative conditions for better test coverage.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #47313

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by leblancd** [comment]
CLA should be all set.

**Comment by cmluciano** [comment]
@k8s-bot ok to test

File: cmd/kube-proxy/app/server_test.go [hunk 1]
```diff
@@ -229,63 +241,90 @@ conntrack:
...
-		},
-		ClusterCIDR:      "1.2.3.0/24",
-		ConfigSyncPeriod: metav1.Duration{Duration: 15 * time.Second},
-		Conntrack: componentconfig.KubeProxyConntrackConfiguration{
-			Max:                   4,
-			MaxPerCore:            2,
-			Min:                   1,
-			TCPCloseWaitTimeout:   metav1.Duration{Duration: 10 * time.Second},
-			TCPEstablishedTimeout: metav1.Duration{Duration: 20 * time.Second},
+	testCases := []struct {
```
Comment (pmichali) [inline cmd/kube-proxy/app/server_test.go]: nit: maybe add a test case name field, and then in the checks latter, can display the test name, so that user knows which test case failed.

File: cmd/kube-proxy/app/server_test.go [hunk 2]
```diff
@@ -134,23 +135,33 @@ func Test_getProxyMode(t *testing.T) {
...
-		t.Fatal(err)
+	// Each bind address below is a separate test case
+	bindAddresses := []string{
+		"0.0.0.0",
+		"2001:db8::1",
 	}
+	for _, addr := range bindAddresses {
+		options, err := NewOptions()
+		if err != nil {
+			t.Fatal(err)
```
Comment (pmichali) [inline cmd/kube-proxy/app/server_test.go]: May be useful to indicate which bind address failed for NewOptions. Also, since working on tests for this source code, it would be a good time to add test cases for failure modes of NewOptions() for better coverage.

File: cmd/kube-proxy/app/server_test.go [hunk 3]
```diff
@@ -229,63 +241,90 @@ conntrack:
...
+		assert.NoError(t, err)
 
-	config, err := options.loadConfig([]byte(yaml))
-	assert.NoError(t, err)
-	if !reflect.DeepEqual(expected, config) {
-		t.Fatalf("unexpected config, diff = %s", diff.ObjectDiff(config, expected))
+		yaml := fmt.Sprintf(
+			yamlTemplate, tc.bindAddress, tc.clusterCIDR,
+			tc.healthzBindAddress, tc.metricsBindAddress)
+		config, err := options.loadConfig([]byte(yaml))
```
Comment (pmichali) [inline cmd/kube-proxy/app/server_test.go]: Good opportunity to add test cases for failure modes of loadConfig() too.

**Review by pmichali — CHANGES_REQUESTED** [review]
Mostly minor improvement suggestions...

**Comment by thockin** [comment]
@pmichali owns this review

I can't /approve until this PR has an associated issue, and the top comment says `fixes #1234`.  can you file an issue for this to close citing the lack of a test.

**Comment by leblancd** [comment]
@thockin : Added issue #.

**Comment by thockin** [comment]
/approve

@pmichali pmichali holds the last lgtm

File: cmd/kube-proxy/app/server_test.go [hunk 4]
```diff
@@ -229,63 +273,120 @@ conntrack:
...
+		}
+
+		options, err := NewOptions()
+		assert.NoError(t, err, "unexpected error for %s: %v", tc.name, err)
+
+		yaml := fmt.Sprintf(
+			yamlTemplate, tc.bindAddress, tc.clusterCIDR,
+			tc.healthzBindAddress, tc.metricsBindAddress)
+		config, err := options.loadConfig([]byte(yaml))
+		assert.NoError(t, err, "unexpected error for %s: %v, tc.name, err")
```
Comment (pmichali) [inline cmd/kube-proxy/app/server_test.go]: Whoops... looks like typo here with position of quote! :)

**Review by pmichali — CHANGES_REQUESTED** [review]
One issue. I didn't check coverage with this change set. Were you able to hit all the error cases?

File: cmd/kube-proxy/app/server_test.go (hunk 4 above)
Comment (leblancd) [inline cmd/kube-proxy/app/server_test.go]: Oops, thanks for catching that.

**Review by leblancd — COMMENTED** [review]
@pmichali Unit tests cover the failure modes for NewOptions and loadConfig as requested.

**Comment by leblancd** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by leblancd** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by pmichali** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
